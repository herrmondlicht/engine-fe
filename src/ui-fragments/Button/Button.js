import React, { useMemo } from "react";
import propTypes from "prop-types";
import Loader from "react-loader-spinner";
import CheckCircleIcon from "@heroicons/react/solid/CheckCircleIcon";
import XCircleIcon from "@heroicons/react/solid/XCircleIcon";
import { theme } from "twtheme";
import { Link } from "react-router-dom";

export const BUTTON_VARIANTS = {
  GHOST: "ghost",
  PRIMARY: "primary",
  SUCCESS: "success",
  ERROR: "error",
  HOLLOW: "hollow",
};

export const BUTTON_SIZES = {
  SMALL: "small",
  BIG: "big",
};

export const BUTTON_COMPONENT = {
  BUTTON: "button",
  LINK: "a",
};

const Button = (
  {
    onClick,
    children,
    variant = "primary",
    size,
    fw,
    disabled,
    showLoader,
    as = BUTTON_COMPONENT.BUTTON,
    href,
    showVariantIcon,
  },
  ref
) => {
  const sizeClasses = useMemo(() => {
    switch (size) {
      case BUTTON_SIZES.SMALL:
        return ["p-2", "px-4", "text-sm"].join(" ");
      case BUTTON_SIZES.BIG:
        return ["p-4", "px-6", "text-lg"].join(" ");
      default:
        return ["p-3", "px-5", "text-md"].join(" ");
    }
  }, [size]);

  const variantClasses = useMemo(() => {
    const variantColor = `${variant}-0`;
    switch (variant) {
      case BUTTON_VARIANTS.GHOST:
        return ["text-primary-0"];
      case BUTTON_VARIANTS.HOLLOW:
        return [
          "bg-transparent",
          "ring-2",
          "ring-primary-0",
          "text-primary-0",
          "hover:text-white",
          "hover:bg-primary-0",
        ].join(" ");
      case BUTTON_VARIANTS.PRIMARY:
      case BUTTON_VARIANTS.SUCCESS:
      case BUTTON_VARIANTS.ERROR:
        return [
          `bg-${variantColor}`,
          `focus:ring-${variantColor}`,
          `hover:bg-${variant}-1`,
          "focus:ring-2",
          "focus:border-transparent",
          "border-2",
          "text-white",
        ].join(" ");
    }
  }, [variant]);

  return (
    <>
      {as === BUTTON_COMPONENT.BUTTON && (
        <button
          disabled={disabled}
          ref={ref}
          onClick={onClick || undefined}
          className={`${sizeClasses} ${variantClasses} ${
            fw ? "w-full" : ""
          } rounded-full font-semibold transition duration-300`}
        >
          <ButtonContent
            showLoader={showLoader}
            variant={variant}
            showVariantIcon={showVariantIcon}
          >
            {children}
          </ButtonContent>
        </button>
      )}
      {as === BUTTON_COMPONENT.LINK && (
        <Link
          disabled={disabled}
          ref={ref}
          to={href}
          className={`${sizeClasses} ${variantClasses} ${
            fw ? "w-full" : ""
          } rounded-full text-white font-semibold cursor-pointer transition duration-300`}
        >
          <ButtonContent showLoader={showLoader} variant={variant}>
            {children}
          </ButtonContent>
        </Link>
      )}
    </>
  );
};

const ButtonContent = ({
  children,
  showLoader,
  variant,
  showVariantIcon = true,
}) => {
  const variantIcon = useMemo(() => {
    const sizeClasses = "h-7 w-7";
    if (!showVariantIcon) {
      return null;
    }
    switch (variant) {
      case "success":
        return <CheckCircleIcon className={sizeClasses} />;
      case "error":
        return <XCircleIcon className={sizeClasses} />;
      default:
        return null;
    }
  }, [showVariantIcon, variant]);
  return (
    <div className="relative w-full h-full flex justify-center items-center">
      <div
        className={`transition duration-400 ${
          showLoader || variant === "success" ? "opacity-0" : "opacity-100"
        }`}
      >
        {children}
      </div>
      <div
        className={`absolute z-10 transition duration-400 ${
          showLoader ? "opacity-100" : "opacity-0"
        }`}
      >
        <Loader
          type="TailSpin"
          height={30}
          width={30}
          color={
            variant === BUTTON_VARIANTS.GHOST
              ? theme.colors.primary[0]
              : theme.colors.white
          }
        />
      </div>
      {!showLoader && (
        <div
          className={`absolute z-10 transition duration-400 ${
            variantIcon ? "opacity-100" : "opacity-0"
          }`}
        >
          {variantIcon}
        </div>
      )}
    </div>
  );
};

const ButtonFR = React.forwardRef(Button);

Button.propTypes = {
  children: propTypes.any,
  onClick: propTypes.func,
  variant: propTypes.string,
  size: propTypes.string,
  fw: propTypes.bool,
  showLoader: propTypes.bool,
  disabled: propTypes.bool,
  showVariantIcon: propTypes.bool,
};
export { ButtonFR as Button };
