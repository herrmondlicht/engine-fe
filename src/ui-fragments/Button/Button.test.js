import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { assert, stub } from "sinon";

import { Button, BUTTON_COMPONENT } from "./Button";

describe("Button", () => {
  it("should render a Button", async () => {
    render(<Button>Text</Button>);
    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("Renders the Button as link ", async () => {
    render(
      <BrowserRouter>
        <Button as={BUTTON_COMPONENT.LINK} href="/">
          Text
        </Button>
      </BrowserRouter>
    );
    expect(screen.getByRole("link")).toMatchSnapshot();
  });

  it("Renders a disabled button ", async () => {
    const onClickStub = stub();
    render(
      <Button onClick={onClickStub} disabled>
        Text
      </Button>
    );
    expect(screen.getByRole("button")).toBeDisabled();
    screen.getByRole("button").click();

    assert.notCalled(onClickStub);
  });

  it("Executes the onClick when clicked", async () => {
    const onClickStub = stub();
    render(<Button onClick={onClickStub}>Text</Button>);
    expect(screen.getByRole("button")).toBeEnabled();
    screen.getByRole("button").click();
    assert.calledOnce(onClickStub);
  });

  it("Renders a button when showLoader is active", async () => {
    render(<Button showLoader>Text</Button>);
    expect(screen.getByRole("button")).toBeDisabled();
    expect(screen.getByRole("button")).toMatchSnapshot();
  });
});
