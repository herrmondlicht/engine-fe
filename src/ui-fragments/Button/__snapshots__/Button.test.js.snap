// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Renders a button when showLoader is active 1`] = `
<button
  class="p-3 px-5 text-md bg-primary-0 focus:ring-primary-0 hover:bg-primary-1 focus:ring-2 focus:border-transparent border-2 text-white  rounded-full font-semibold transition duration-300"
  disabled=""
>
  <div
    class="relative w-full h-full flex justify-center items-center"
  >
    <div
      class="transition duration-400 opacity-0"
    >
      Text
    </div>
    <div
      class="absolute z-10 transition duration-400 opacity-100"
    >
      <div
        aria-busy="true"
        class=""
      >
        <svg
          aria-label="audio-loading"
          height="30"
          viewBox="0 0 38 38"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <lineargradient
              id="a"
              x1="8.042%"
              x2="65.682%"
              y1="0%"
              y2="23.865%"
            >
              <stop
                offset="0%"
                stop-color="#fff"
                stop-opacity="0"
              />
              <stop
                offset="63.146%"
                stop-color="#fff"
                stop-opacity=".631"
              />
              <stop
                offset="100%"
                stop-color="#fff"
              />
            </lineargradient>
          </defs>
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <g
              transform="translate(1 1)"
            >
              <path
                d="M36 18c0-9.94-8.06-18-18-18"
                id="Oval-2"
                stroke="#fff"
                stroke-width="2"
              >
                <animatetransform
                  attributeName="transform"
                  dur="0.9s"
                  from="0 18 18"
                  repeatCount="indefinite"
                  to="360 18 18"
                  type="rotate"
                />
              </path>
              <circle
                cx="36"
                cy="18"
                fill="#fff"
                r="1"
              >
                <animatetransform
                  attributeName="transform"
                  dur="0.9s"
                  from="0 18 18"
                  repeatCount="indefinite"
                  to="360 18 18"
                  type="rotate"
                />
              </circle>
            </g>
          </g>
        </svg>
      </div>
    </div>
  </div>
</button>
`;

exports[`Button Renders the Button as link  1`] = `
<a
  class="p-3 px-5 text-md bg-primary-0 focus:ring-primary-0 hover:bg-primary-1 focus:ring-2 focus:border-transparent border-2 text-white  rounded-full text-white font-semibold cursor-pointer transition duration-300"
  href="/"
>
  <div
    class="relative w-full h-full flex justify-center items-center"
  >
    <div
      class="transition duration-400 opacity-100"
    >
      Text
    </div>
    <div
      class="absolute z-10 transition duration-400 opacity-0"
    >
      <div
        aria-busy="true"
        class=""
      >
        <svg
          aria-label="audio-loading"
          height="30"
          viewBox="0 0 38 38"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <lineargradient
              id="a"
              x1="8.042%"
              x2="65.682%"
              y1="0%"
              y2="23.865%"
            >
              <stop
                offset="0%"
                stop-color="#fff"
                stop-opacity="0"
              />
              <stop
                offset="63.146%"
                stop-color="#fff"
                stop-opacity=".631"
              />
              <stop
                offset="100%"
                stop-color="#fff"
              />
            </lineargradient>
          </defs>
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <g
              transform="translate(1 1)"
            >
              <path
                d="M36 18c0-9.94-8.06-18-18-18"
                id="Oval-2"
                stroke="#fff"
                stroke-width="2"
              >
                <animatetransform
                  attributeName="transform"
                  dur="0.9s"
                  from="0 18 18"
                  repeatCount="indefinite"
                  to="360 18 18"
                  type="rotate"
                />
              </path>
              <circle
                cx="36"
                cy="18"
                fill="#fff"
                r="1"
              >
                <animatetransform
                  attributeName="transform"
                  dur="0.9s"
                  from="0 18 18"
                  repeatCount="indefinite"
                  to="360 18 18"
                  type="rotate"
                />
              </circle>
            </g>
          </g>
        </svg>
      </div>
    </div>
    <div
      class="absolute z-10 transition duration-400 opacity-0"
    />
  </div>
</a>
`;

exports[`Button should render a Button 1`] = `
<button
  class="p-3 px-5 text-md bg-primary-0 focus:ring-primary-0 hover:bg-primary-1 focus:ring-2 focus:border-transparent border-2 text-white  rounded-full font-semibold transition duration-300"
>
  <div
    class="relative w-full h-full flex justify-center items-center"
  >
    <div
      class="transition duration-400 opacity-100"
    >
      Text
    </div>
    <div
      class="absolute z-10 transition duration-400 opacity-0"
    >
      <div
        aria-busy="true"
        class=""
      >
        <svg
          aria-label="audio-loading"
          height="30"
          viewBox="0 0 38 38"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <lineargradient
              id="a"
              x1="8.042%"
              x2="65.682%"
              y1="0%"
              y2="23.865%"
            >
              <stop
                offset="0%"
                stop-color="#fff"
                stop-opacity="0"
              />
              <stop
                offset="63.146%"
                stop-color="#fff"
                stop-opacity=".631"
              />
              <stop
                offset="100%"
                stop-color="#fff"
              />
            </lineargradient>
          </defs>
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <g
              transform="translate(1 1)"
            >
              <path
                d="M36 18c0-9.94-8.06-18-18-18"
                id="Oval-2"
                stroke="#fff"
                stroke-width="2"
              >
                <animatetransform
                  attributeName="transform"
                  dur="0.9s"
                  from="0 18 18"
                  repeatCount="indefinite"
                  to="360 18 18"
                  type="rotate"
                />
              </path>
              <circle
                cx="36"
                cy="18"
                fill="#fff"
                r="1"
              >
                <animatetransform
                  attributeName="transform"
                  dur="0.9s"
                  from="0 18 18"
                  repeatCount="indefinite"
                  to="360 18 18"
                  type="rotate"
                />
              </circle>
            </g>
          </g>
        </svg>
      </div>
    </div>
    <div
      class="absolute z-10 transition duration-400 opacity-0"
    />
  </div>
</button>
`;
